### PR TITLE
feat(renderers): add `MetadataRegistry` to set custom party card metadata

### PR DESCRIPTION
**Pull request by 0xble**

### CODE REVIEW

File: contracts/renderers/MetadataRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    error OnlyAllowedHostsError();
+    error OnlyMultisigError();
+    error InvalidCrowdfundError();
+
+    Globals private immutable _GLOBALS;
+
+    /// @notice The address of the host that is allowed to set custom token URIs
+    mapping(address => bool) public allowedHosts;
+    /// @notice The custom token URI metadata for the party created by a given crowdfund
+    mapping(Crowdfund => TokenMetadata) public customPartyTokenURIByCrowdfund;
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:32]: feels like this and `customPartyContractURIByCrowdfund` should be renamed like `customPartyMetadataByCrowdfund`

File: contracts/renderers/MetadataRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    modifier onlyMultisig() {
+        if (msg.sender != _GLOBALS.multiSig()) revert OnlyMultisigError();
+        _;
+    }
+
+    /// @notice Set the status of a host as allowed to set custom token URIs or not
+    /// @param host The host to set the status of
+    /// @param allowed Whether the host is allowed or not
+    function setAllowedHosts(address host, bool allowed) external onlyMultisig {
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:57]: should add event

File: contracts/renderers/MetadataRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        allowedHosts[host] = allowed;
+    }
+
+    /// @notice Set the custom token URI for a party created by a given crowdfund
+    /// @param crowdfund The crowdfund that created the party
+    /// @param metadata The metadata for the custom token URI to set for the party's token
+    function setCustomTokenURI(
+        Crowdfund crowdfund,
+        TokenMetadata memory metadata
+    ) external onlyAllowedHosts {
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:67]: should add event

File: contracts/renderers/MetadataRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    /// @param host The host to set the status of
+    /// @param allowed Whether the host is allowed or not
+    function setAllowedHosts(address host, bool allowed) external onlyMultisig {
+        allowedHosts[host] = allowed;
+    }
+
+    /// @notice Set the custom token URI for a party created by a given crowdfund
+    /// @param crowdfund The crowdfund that created the party
+    /// @param metadata The metadata for the custom token URI to set for the party's token
+    function setCustomTokenURI(
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:64]: should be named like `setCustomTokenMetadata` imho

File: contracts/renderers/MetadataRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            revert MetadataAlreadySetError();
+        }
+
+        customPartyTokenURIByCrowdfund[crowdfund] = metadata;
+    }
+
+    /// @notice Set the custom contract URI for a party created by a given crowdfund
+    /// @param crowdfund The crowdfund that created the party
+    /// @param metadata The metadata for the custom contract URI to set for the party's contract
+    function setCustomContractURI(
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:88]: should be named like `setCustomContractMetadata` imho

File: contracts/renderers/MetadataRegistry.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        Crowdfund crowdfund,
+        ContractMetadata memory metadata
+    ) external onlyAllowedHosts {
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:91]: should add event

File: contracts/renderers/MetadataRegistry.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        if (address(crowdfund) == address(0)) {
+            revert InvalidCrowdfundError();
+        }
+
+        // Prevent setting the metadata if it has already been set
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:96]: per convo in discord, we could remove this check, and instead check:
- is crowdfund balance zero
- is crowdfund status active

File: contracts/renderers/MetadataRegistry.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    constructor(Globals global, address[] memory _allowedHosts) {
+        _GLOBALS = global;
+
+        for (uint256 i; i < _allowedHosts.length; i++) {
+            allowedHosts[_allowedHosts[i]] = true;
+        }
+    }
+
+    modifier onlyAllowedHosts() {
```
Comment (steveklebanoff) [inline contracts/renderers/MetadataRegistry.sol:44]: could consider allowing multisig to allow anyone to set custom metadata, but this is optional

**Review by steveklebanoff — COMMENTED** [review]
added a few suggestions

File: contracts/renderers/PartyNFTRenderer.sol [hunk 9]
```diff
@@ -38,11 +39,118 @@ contract PartyNFTRenderer is RendererBase {
...
+                        "data:application/json;base64,",
+                        Base64.encode(
+                            abi.encodePacked(
+                                '{"name":"',
+                                name,
+                                '", "description":"',
+                                description,
+                                '", "image":"',
+                                image,
+                                '", "banner":"',
```
Comment (steveklebanoff) [inline contracts/renderers/PartyNFTRenderer.sol:77]: Can we include the auto external_url of the partybid address here?

File: contracts/renderers/CrowdfundNFTRenderer.sol [hunk 10]
```diff
@@ -29,6 +29,34 @@ contract CrowdfundNFTRenderer is RendererBase {
...
+            string.concat(
+                "data:application/json;base64,",
+                Base64.encode(
+                    abi.encodePacked(
+                        '{"name":"',
+                        generateCollectionName(),
+                        '", "description":"',
+                        generateCollectionDescription(),
+                        '", "external_url":"',
+                        "https://www.partybid.app/"
```
Comment (steveklebanoff) [inline contracts/renderers/CrowdfundNFTRenderer.sol:49]: can we use the URL of the specific crowdfund?

File: contracts/renderers/PartyNFTRenderer.sol [hunk 11]
```diff
@@ -38,11 +39,118 @@ contract PartyNFTRenderer is RendererBase {
...
+            string.concat(
+                "data:application/json;base64,",
+                Base64.encode(
+                    abi.encodePacked(
+                        '{"name":"',
+                        generateCollectionName(),
+                        '", "description":"',
+                        generateCollectionDescription(),
+                        '", "external_url":"',
+                        "https://www.partybid.app/"
```
Comment (steveklebanoff) [inline contracts/renderers/PartyNFTRenderer.sol:102]: can we link to the specific party here?

File: contracts/renderers/PartyNFTRenderer.sol [hunk 12]
```diff
@@ -38,11 +39,118 @@ contract PartyNFTRenderer is RendererBase {
...
+                bytes(image).length > 0 &&
+                bytes(banner).length > 0
+            ) {
+                return
+                    string.concat(
+                        "data:application/json;base64,",
+                        Base64.encode(
+                            abi.encodePacked(
+                                '{"name":"',
+                                name,
```
Comment (steveklebanoff) [inline contracts/renderers/PartyNFTRenderer.sol:72]: Can we include the item id after the name? i.e. `[custom name] #[item id]`

Comment (0xble) [reply contracts/renderers/PartyNFTRenderer.sol:72]: This is for the collection name. I assume you mean for the party card name?

Comment (steveklebanoff) [reply contracts/renderers/PartyNFTRenderer.sol]: ah, yes -- my b

File: sol-tests/party/PartyGovernanceNFT.t.sol [hunk 13]
```diff
@@ -398,6 +468,33 @@ contract PartyGovernanceNFTTest is Test, TestUtils {
...
 
+contract DummyCrowdfund {
+    uint256 public totalContributions;
+    bytes32 public governanceOptsHash;
+
+    function setGovernanceOptsHash(Crowdfund.FixedGovernanceOpts memory opts) external {
+        governanceOptsHash = _hashFixedGovernanceOpts(opts);
+    }
+
+    function _hashFixedGovernanceOpts(
```
Comment (horsefacts) [inline sol-tests/party/PartyGovernanceNFT.t.sol:479]: Nitpick, but I hate seeing this copy-pasted all over. Maybe a use case for a free function?

File: contracts/renderers/CrowdfundNFTRenderer.sol [hunk 14]
```diff
@@ -29,6 +29,34 @@ contract CrowdfundNFTRenderer is RendererBase {
...
+        return
+            string.concat(
+                "data:application/json;base64,",
+                Base64.encode(
+                    abi.encodePacked(
+                        '{"name":"',
+                        generateCollectionName(),
+                        '", "description":"',
+                        generateCollectionDescription(),
+                        '", "external_url":"',
```
Comment (horsefacts) [inline contracts/renderers/CrowdfundNFTRenderer.sol:48]: This is called `external_link` in the OpenSea docs [here](https://docs.opensea.io/docs/contract-level-metadata), but I would not be at all surprised if you are right and the docs are wrong, or if both are actually supported.

**Pull request merged**